barman/server.py

```diff
@@ -1516,47 +1516,85 @@ def delete_backup(self, backup):
+                "Backup %s has incremental backups which depend on it. "
+                "Deleting all backups in the tree",
+                backup.backup_id,
+            )
+
+        try:
+            # Lock acquisition: if you can acquire a ServerBackupLock it means
+            # that no other processes like a backup or another delete is running
+            with ServerBackupLock(self.config.barman_lock_directory, self.config.name):
+                # Delete the backup along with all its descendants in the
+                # backup tree i.e. all its subsequent incremental backups.
+                # If it has no descendants or it is an rsync backup then
+                # only the current backup is deleted.
+                deleted = False
+                backups_to_delete = backup.walk_backups_tree()
+                for del_backup in backups_to_delete:
+                    deleted = self.perform_delete_backup(del_backup)
+                    if not deleted and del_backup.backup_id != backup.backup_id:
+                        output.error(
+                            "Failed to delete one of its incremental backups. Make sure "
+                            "all its dependent backups are deletable and try again."
+                        )
+                        break
+
+                return deleted
 
         except LockFileBusy:
-            # Otherwise if the lockfile is busy, a backup process is actually
-            # running on that server. To be sure that it's safe
-            # to delete the backup, we must check its status and its position
-            # in the catalogue.
-            # If it is the first and it is STARTED or EMPTY, we are trying to
-            # remove a running backup. This operation must be forbidden.
-            # Otherwise, normally delete the backup.
-            first_backup_id = self.get_first_backup_id(BackupInfo.STATUS_ALL)
-            if backup.backup_id == first_backup_id and backup.status in (
-                BackupInfo.STARTED,
-                BackupInfo.EMPTY,
-            ):
-                output.error(
-                    "Another action is in progress for the backup %s"
-                    " of server %s. Impossible to delete the backup."
-                    % (backup.backup_id, self.config.name)
-                )
-                return
+            # Otherwise if the lockfile is busy, a backup process is actually running
+            output.error(
+                "Another process in running on server %s. "
+                "Impossible to delete the backup." % self.config.name
+            )
+            return False
 
         except LockFilePermissionDenied as e:
             # We cannot access the lockfile.
             # Exit without removing the backup.
             output.error("Permission denied, unable to access '%s'" % e)
-            return
+            return False
 
+    def perform_delete_backup(self, backup):
+        """Perform the deletion of a backup
```

Comment:
I'd say to describe in this docstring what this method does so we can get the difference between `delete_backup` and `perform_delete_backup` in the future, if needed.

I mean, something similar to what you did in your commit, which is clearer.